leader/undirected_ring/sync_hirschberg_sinclair_second_edition/sync_hirschberg_sinclair_second_edition.go

```diff
@@ -0,0 +1,301 @@
+	var jsonReceived []byte
+	var replyFromL bool
+	var LReplyed bool
+	var RReplyed bool
+	var receivedData SendingData
+
+	dataToSend := cand.InitSendingData(From, cand.myValue, 0, maxnum)
+	cand.Sendboth(dataToSend)
+	counter := 0
+	final := false
+	for {
+		select {
+		case jsonReceived = <-cand.leftInput:
+			replyFromL = true
+		case jsonReceived = <-cand.rightInput:
+			replyFromL = false
+		}
+		_ = json.Unmarshal(jsonReceived, &receivedData)
+		switch receivedData.MessageType {
+		case From:
+			cand.ReceivedMessageIsFrom(receivedData, replyFromL)
+			if cand.myStatus == WonStatus {
+				counter++
+				if counter == 2 {
+					dataToSend.MessageType = End
+					cand.Sendpass(&dataToSend, replyFromL)
+				}
+			}
+
+		case No, Ok:
+			cand.ReceivedMessageIsOkOrNo(receivedData, replyFromL)
+			if receivedData.MessageType == No {
+				cand.myStatus = LostStatus
+			}
+			if receivedData.Value == cand.myValue {
+				if replyFromL {
+					LReplyed = true
+				} else {
+					RReplyed = true
+				}
+				if LReplyed && RReplyed && cand.myStatus == CandidateStatus {
+					LReplyed = false
+					RReplyed = false
+					maxnum *= 2
+					dataToSend := cand.InitSendingData(From, cand.myValue, 0, maxnum)
+					cand.Sendboth(dataToSend)
+				}
+			}
+		case End:
+			if cand.myStatus == LostStatus {
+				cand.Sendpass(&receivedData, replyFromL)
+			}
+			final = true
+
+		default:
+			//nothing to do
+		}
+
+		if final {
+			break
+		}
+	}
+
+	if cand.myStatus == WonStatus {
+		return cand.myValue
+	} else {
+		return -1
+	}
+}
+
+func NewCandidate(
+	myValue int,
+	leftInput <-chan []byte,
+	leftOutput chan<- []byte,
+	rightInput <-chan []byte,
+	rightOutput chan<- []byte) ICandidate {
+	return &Candidate{
+		myValue:     myValue,
+		myStatus:    CandidateStatus,
+		leftInput:   leftInput,
+		leftOutput:  leftOutput,
+		rightInput:  rightInput,
+		rightOutput: rightOutput,
+	}
+}
+
+func (cand *Candidate) sendTo(channel chan<- []byte, jsonToSend []byte) {
+	go func() { channel <- jsonToSend }()
+}
+
+func (cand *Candidate) InitSendingData(messageType MessageType, value int, num uint64, maxnum uint64) SendingData {
+	return SendingData{
+		MessageType: messageType,
+		Value:       value,
+		Num:         num,
+		Maxnum:      maxnum,
+	}
+}
+
+func BuildUndirectedRingAndRun(N int, argsArr []int) {
+	firstInput1 := make(chan []byte)
+	firstOutput1 := make(chan []byte)
+
+	prevOutput := firstInput1
+	prevInput := firstOutput1
+
+	writeChannel := make(chan int)
+	var wg sync.WaitGroup
+	wg.Add(N)
+	runner := func(candidate ICandidate) {
+		defer wg.Done()
+		leader := candidate.SelectLeader()
+		writeChannel <- leader
+	}
+
+	//arr := generateShuffledArray(N)
+	var arrBits []bool
+
+	for it := 0; it < N; it++ {
+		nextOutput := make(chan []byte)
+		nextInput := make(chan []byte)
+
+		var input1, input2 <-chan []byte
+		var output1, output2 chan<- []byte
+		input1 = prevOutput
+		output1 = prevInput
+		input2 = nextInput
+		output2 = nextOutput
```

Comment:
`input1, input2 := prevOutput, nextInput`

Here it's irrelevant since the whole function is to be removed, but please apply the shorthand notation elsewhere

Reply:
Has been removed